### PR TITLE
App class and scenes

### PR DESCRIPTION
**Pull request by MisterBlue1369**
implement of app class to separe it from main function

and implement scene class for the gameplay scene and later use it for menus, etc

### CODE REVIEW

File: headers/Scene.hpp [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+class Application;
+
+class Scene {
+	protected:
+		Application* app;
+	public:
+		Scene(Application* app);
+		virtual void OnCreate() = 0;
+		virtual void OnUpdate() = 0;
```
Comment (sjtolley9) [inline headers/Scene.hpp]: We'll want to pass `elapsedTime` to the update method so updates are based on time and not frames.
```suggestion
		virtual void OnUpdate(Uint32 elapsedTime) = 0;
```

File: headers/Application.hpp [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+#define GHOSTS 896*20
+
+class Application {
+	public:
+		Application();
+		void init();
+		void eventHandling();
+		void render();
+		void update();
+		void run();
```
Comment (sjtolley9) [inline headers/Application.hpp:26]: We should include methods to push/pop methods from the Scene stack, allowing gameplay to be passed to other scenes. Make sure we update `currentScene` when we do this.

**Review by sjtolley9 — COMMENTED** [review]
There's a couple changes requested to improve the design, but overall the structure looks really good!

**Comment by MisterBlue1369** [comment]
i think is done

File: headers/Scene.hpp [hunk 3]
```diff
@@ -8,7 +10,7 @@ class Scene {
 	public:
 		Scene(Application* app);
 		virtual void OnCreate() = 0;
-		virtual void OnUpdate() = 0;
+		virtual void OnUpdate(uint32_t elapsedTime) = 0;
```
Comment (sjtolley9) [inline headers/Scene.hpp]: This should use the `Uint32` type from `"SDL.h"` so as to keep types consistent throughout.

File: src/Application.cpp [hunk 4]
```diff
@@ -55,7 +56,24 @@ void Application::render () {
...
+	scenes[currentScene]->OnUpdate(elapsedTime);
+}
+
+void Application::changeSceneTo(int sceneIndex){
+	if(sceneIndex < 0 || sceneIndex >= scenes.size()) {
+		std::cout<<"scene index non-existend\n";
+		return;
+	}
+	currentScene = sceneIndex;
+}
```
Comment (sjtolley9) [inline src/Application.cpp]: I'm not sure we need this method. Push and pop should be enough. Plus relying on `sceneIndex` for changing could lead to unexpected results, see next comment. Removing this and the `removeScene` will allow the scene container to better reflect a stack.

File: src/Application.cpp [hunk 5]
```diff
@@ -55,7 +56,24 @@ void Application::render () {
...
+void Application::pushBackScene(Scene& scene){
+	scenes.push_back(&scene);
+}
+void Application::removeScene(int sceneIndex){
+	scenes.erase(scenes.begin()+sceneIndex);
+}
```
Comment (sjtolley9) [inline src/Application.cpp]: As I mentioned, using a `sceneIndex` to change or remove scenes could result in weird behavior that would be difficult to debug. An example of this being that if we were to remove a scene, the scene index for all scenes placed on the stack after that scene will decrease by one.

**Review by sjtolley9 — CHANGES_REQUESTED** [review]

**Comment by sjtolley9** [comment]
### An important note:
After this Pull Request gets merged, I'm going to be adding in a few new things and won't be merging in pull requests unless they directly address an existing issue.

**Comment by MisterBlue1369** [comment]
now i think is ready?

**Review by sjtolley9 — COMMENTED** [review]
Looks good, I'll do a quick compile check and merge it in.

**Review by sjtolley9 — APPROVED** [review]

**Pull request merged**